File: chart/src/main/java/cn/jingzhuan/lib/chart3/renderer/AxisRenderer.java
package cn.jingzhuan.lib.chart3.renderer;

import android.graphics.Canvas;
import android.graphics.DashPathEffect;
import android.graphics.Paint;

import java.util.List;

import cn.jingzhuan.lib.chart.AxisAutoValues;
import cn.jingzhuan.lib.chart.component.Axis;
import cn.jingzhuan.lib.chart.component.AxisX;
import cn.jingzhuan.lib.chart.component.AxisY;
import cn.jingzhuan.lib.chart.data.AbstractDataSet;
import cn.jingzhuan.lib.chart.data.LabelColorSetter;
import cn.jingzhuan.lib.chart.data.Value;
import cn.jingzhuan.lib.chart3.base.AbstractChartView;

/**
 * @since 2023-09-04
 * created by lei
 */
public class AxisRenderer<V extends Value, T extends AbstractDataSet<V>> extends AbstractRenderer<V, T> {

    private final Axis mAxis;

    private Paint mAxisPaint;

    private Paint mGridPaint;

    private final char[] mLabelBuffer = new char[100];

    public AxisRenderer(AbstractChartView<V, T> chart, Axis axis) {
        super(chart);
        this.mAxis = axis;
        initPaints();
    }

    private void initPaints() {
        mAxisPaint = new Paint();
        mAxisPaint.setStyle(Paint.Style.STROKE);

        mGridPaint = new Paint();
        mGridPaint.setStyle(Paint.Style.STROKE);

        mLabelTextPaint.setTextSize(mAxis.getLabelTextSize());
        mLabelTextPaint.setColor(mAxis.getLabelTextColor());
        mAxis.setLabelWidth((int) mLabelTextPaint.measureText("0000"));
        if (mAxis.getLabelTextSize() > 0) {
            mAxis.setLabelHeight((int) Math.abs(mLabelTextPaint.getFontMetrics().top));
        }
    }

    /**
     * 画坐标轴
     */
    @Override
    public void renderer(Canvas canvas) {
        if (mAxis.getLabels() == null) {
            if (mAxis instanceof AxisX) {
                computeAxisStopsX(mViewport.left, mViewport.right, (AxisX) mAxis, null);
            } else if (mAxis instanceof AxisY) {
                computeAxisStopsY((AxisY) mAxis);
            }
        }
        // Draws lib container
        drawAxisLine(canvas);
    }

    private void drawAxisLine(Canvas canvas) {
        float startX = 0f, startY = 0f, stopX = 0f, stopY = 0f;
        float halfThickness = mAxis.getAxisThickness() * .5f;
        switch (mAxis.getAxisPosition()) {
            case AxisX.TOP:
            case AxisX.TOP_INSIDE:
                startX = mContentRect.left;
                startY = mContentRect.top + halfThickness;
                stopX = mContentRect.right;
                stopY = startY;
                break;
            case AxisX.BOTTOM:
            case AxisX.BOTTOM_INSIDE:
                startX = mContentRect.left;
                startY = mContentRect.bottom - halfThickness;
                stopX = mContentRect.right;
                stopY = startY;
                break;
            case AxisY.LEFT_INSIDE:
            case AxisY.LEFT_OUTSIDE:
                startX = mContentRect.left + halfThickness;
                startY = mContentRect.top;
                stopX = startX;
                stopY = mContentRect.bottom;
                break;
            case AxisY.RIGHT_INSIDE:
            case AxisY.RIGHT_OUTSIDE:
                startX = mContentRect.right - halfThickness;
                startY = mContentRect.top;
                stopX = startX;
                stopY = mContentRect.bottom;
                break;
        }
        // Draw axis line
        if (mAxis.isEnable()) {
            mAxisPaint.setStrokeWidth(mAxis.getAxisThickness());
            mAxisPaint.setColor(mAxis.getAxisColor());
            canvas.drawLine(startX, startY, stopX, stopY, mAxisPaint);
        }
    }

    private void computeAxisStopsX(float start, float stop, AxisX axis, AxisAutoValues autoValues) {
        double range = stop - start;
        if (axis.getGridCount() == 0 || range <= 0) {
            return;
        }

        final int count = axis.getGridCount() + 1;
        double rawInterval = range / count;

        double interval = roundToOneSignificantFigure(rawInterval);
        double first = Math.ceil(start / interval) * interval;

        double f;
        int i;

        axis.mLabelEntries = new float[count + 1];
        for (f = first, i = 0; i < count + 1; f += interval, ++i) {
            axis.mLabelEntries[i] = (float) f;
        }

    }

    /**
     * Rounds the given number to the given number of significant digits. Based on an answer on
     * <a href="http://stackoverflow.com/questions/202302">Stack Overflow</a>.
     */
    private float roundToOneSignificantFigure(double num) {
        final float d = (float) Math.ceil((float) Math.log10(num < 0 ? -num : num));
        final int power = 1 - (int) d;
        final float magnitude = (float) Math.pow(10, power);
        final long shifted = Math.round(num * magnitude);
        return shifted / magnitude;
    }

    private void computeAxisStopsY(AxisY axis) {
        double min = axis.getYMin();
        double max = axis.getYMax();
        int count = axis.getGridCount() + 1;

        double interval = (max - min) / count;

        axis.mLabelEntries = new float[count + 1];

        if (min < Float.MAX_VALUE && max > -Float.MAX_VALUE && min <= max) {
            double f = min;
            for (int j = 0; j < count + 1; f += interval, j++) {
                axis.mLabelEntries[j] = (float) f;
            }
        }
    }

    /**
     * 画坐标轴文本
     */
    public void drawAxisLabels(Canvas canvas) {
        if (!mAxis.isLabelEnable()) return;
        if (mAxis.getLabels() == null || mAxis.getLabels().isEmpty()) return;

        List<String> labels = mAxis.getLabels();

        mLabelTextPaint.setColor(mAxis.getLabelTextColor());
        mLabelTextPaint.setTextSize(mAxis.getLabelTextSize());

        float x = 0f, y = 0f;
        switch (mAxis.getAxisPosition()) {
            case AxisX.TOP:
            case AxisX.TOP_INSIDE:
                x = mContentRect.left;
                y = mContentRect.top;
                break;
            case AxisX.BOTTOM:
            case AxisX.BOTTOM_INSIDE:
                x = mContentRect.left;
                y = mContentRect.top + mContentRect.height();
                break;
            case AxisY.LEFT_INSIDE:
            case AxisY.LEFT_OUTSIDE:
                x = mContentRect.left;
                y = mContentRect.bottom;
                break;
            case AxisY.RIGHT_INSIDE:
            case AxisY.RIGHT_OUTSIDE:
                x = mContentRect.right;
                y = mContentRect.bottom;
                break;
        }
        int labelOffset;
        int labelLength;

        if (mAxis instanceof AxisX) {
            // X轴
            final float width = mContentRect.width() / ((float) labels.size());
            for (int i = 0; i < labels.size(); i++) {
                char[] labelCharArray = labels.get(i).toCharArray();
                labelLength = labelCharArray.length;
                System.arraycopy(labelCharArray, 0, mLabelBuffer, mLabelBuffer.length - labelLength, labelLength);

                labelOffset = mLabelBuffer.length - labelLength;

                mLabelTextPaint.setTextAlign(Paint.Align.CENTER);

                LabelColorSetter colorSetter = mAxis.getLabelColorSetter();
                if (colorSetter != null) {
                    mLabelTextPaint.setColor(colorSetter.getColorByIndex(i));
                }

                canvas.drawText(mLabelBuffer, labelOffset, labelLength,
                        width * 0.5f + getDrawX(i / ((float) labels.size())),
                        y + mLabelTextPaint.getTextSize(), mLabelTextPaint);
            }
        } else {
            // Y轴
            final float height = mContentRect.height() / (mAxis.getLabels().size() - 1F);
            float separation = 0;
            for (int i = 0; i < mAxis.getLabels().size(); i++) {
                char[] labelCharArray = mAxis.getLabels().get(i).toCharArray();
                labelLength = labelCharArray.length;
                System.arraycopy(labelCharArray, 0, mLabelBuffer, mLabelBuffer.length - labelLength, labelLength);

                labelOffset = mLabelBuffer.length - labelLength;
                switch (mAxis.getAxisPosition()) {
                    case AxisY.LEFT_OUTSIDE:
                    case AxisY.RIGHT_INSIDE:
                        mLabelTextPaint.setTextAlign(Paint.Align.RIGHT);
                        separation = -mAxis.getLabelSeparation();
                        break;
                    case AxisY.LEFT_INSIDE:
                    case AxisY.RIGHT_OUTSIDE:
                        mLabelTextPaint.setTextAlign(Paint.Align.LEFT);
                        separation = mAxis.getLabelSeparation();
                        break;
                }

                float textHeightOffset = (mLabelTextPaint.descent() + mLabelTextPaint.ascent()) / 2;

                LabelColorSetter colorSetter = mAxis.getLabelColorSetter();
                if (colorSetter != null) {
                    mLabelTextPaint.setColor(colorSetter.getColorByIndex(i));
                }

                canvas.drawText(mLabelBuffer, labelOffset, labelLength,
                        x + separation,
                        y - i * height - textHeightOffset,
                        mLabelTextPaint);
            }
        }
    }

    /**
     * 画网格线
     */
    public void drawGridLines(Canvas canvas) {
        int count = mAxis.getGridCount() + 1;

        mGridPaint.setStrokeWidth(mAxis.getGridThickness());
        mGridPaint.setColor(mAxis.getGridColor());

        if (mAxis.getDashedGridIntervals() != null && mAxis.getDashedGridPhase() > 0) {
            mGridPaint.setPathEffect(new DashPathEffect(mAxis.getDashedGridIntervals(), mAxis.getDashedGridPhase()));
        }

        if (mAxis instanceof AxisX) {
            final float width = mContentRect.width() / ((float) count);
            for (int i = 1; i < count; i++) {
                if (mAxis.getGirdLineColorSetter() != null) {
                    mGridPaint.setColor(mAxis.getGirdLineColorSetter().getColorByIndex(mAxis.getGridColor(), i));
                }
                canvas.drawLine(
                        mContentRect.left + i * width,
                        mContentRect.top,
                        mContentRect.left + i * width,
                        mContentRect.bottom,
                        mGridPaint);
            }
        }
        if (mAxis instanceof AxisY) {
            final float height = mContentRect.height() / ((float) count);
            for (int i = 1; i < count; i++) {
                if (mAxis.getGirdLineColorSetter() != null) {
                    mGridPaint.setColor(mAxis.getGirdLineColorSetter().getColorByIndex(mAxis.getGridColor(), i));
                }
                canvas.drawLine(
                        mContentRect.left,
                        mContentRect.top + i * height,
                        mContentRect.right,
                        mContentRect.top + i * height,
                        mGridPaint);
            }
        }
    }

    public Axis getAxis() {
        return this.mAxis;
    }
}

File: chart/src/main/java/cn/jingzhuan/lib/chart3/renderer/AxisRenderer.kt
package cn.jingzhuan.lib.chart3.renderer

import android.graphics.Canvas
import android.graphics.DashPathEffect
import android.graphics.Paint
import cn.jingzhuan.lib.chart.utils.FloatUtils
import cn.jingzhuan.lib.chart3.axis.Axis
import cn.jingzhuan.lib.chart3.axis.AxisX
import cn.jingzhuan.lib.chart3.axis.AxisY
import cn.jingzhuan.lib.chart3.base.AbstractChartView
import cn.jingzhuan.lib.chart3.data.ChartData
import cn.jingzhuan.lib.chart3.data.dataset.AbstractDataSet
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.log10
import kotlin.math.pow
import kotlin.math.roundToInt

/**
 * @since 2023-09-05
 * created by lei
 */
class AxisRenderer<T : AbstractDataSet<*>>(
    chart: AbstractChartView<T>,
    val axis: Axis,
) : AbstractRenderer<T>(chart) {

    private lateinit var axisPaint: Paint

    private lateinit var gridPaint: Paint

    private val labelBuffer = CharArray(100)

    init {
        initPaints()
    }

    private fun initPaints() {
        axisPaint = Paint()
        axisPaint.style = Paint.Style.STROKE

        gridPaint = Paint()
        gridPaint.style = Paint.Style.STROKE

        labelTextPaint.textSize = axis.labelTextSize
        labelTextPaint.color = axis.labelTextColor

        axis.labelWidth = labelTextPaint.measureText("0000").toInt()
        if (axis.labelTextSize > 0) {
            axis.labelHeight = abs(labelTextPaint.fontMetrics.top).toInt()
        }
    }

    /**
     * 画坐标轴
     */
    override fun renderer(canvas: Canvas) {
        if (axis.labels == null) {
            if (axis is AxisX) {
                computeAxisStopsX(currentViewport.left, currentViewport.right, axis)
            } else if (axis is AxisY) {
                computeAxisStopsY(axis)
            }
        }
        // Draws lib container
        drawAxisLine(canvas)
    }

    private fun drawAxisLine(canvas: Canvas) {
        var startX = 0f
        var startY = 0f
        var stopX = 0f
        var stopY = 0f
        val halfThickness = axis.axisThickness * .5f
        when (axis.axisPosition) {
            AxisX.TOP, AxisX.TOP_INSIDE -> {
                startX = contentRect.left.toFloat()
                startY = contentRect.top + halfThickness
                stopX = contentRect.right.toFloat()
                stopY = startY
            }

            AxisX.BOTTOM, AxisX.BOTTOM_INSIDE -> {
                startX = contentRect.left.toFloat()
                startY = contentRect.bottom - halfThickness
                stopX = contentRect.right.toFloat()
                stopY = startY
            }

            AxisY.LEFT_INSIDE, AxisY.LEFT_OUTSIDE -> {
                startX = contentRect.left + halfThickness
                startY = contentRect.top.toFloat()
                stopX = startX
                stopY = contentRect.bottom.toFloat()
            }

            AxisY.RIGHT_INSIDE, AxisY.RIGHT_OUTSIDE -> {
                startX = contentRect.right - halfThickness
                startY = contentRect.top.toFloat()
                stopX = startX
                stopY = contentRect.bottom.toFloat()
            }
        }
        // Draw axis line
        if (axis.isEnable) {
            axisPaint.strokeWidth = axis.axisThickness
            axisPaint.color = axis.axisColor
            canvas.drawLine(startX, startY, stopX, stopY, axisPaint)
        }
    }

    private fun computeAxisStopsX(
        start: Float,
        stop: Float,
        axis: AxisX,
    ) {
        val range = (stop - start).toDouble()
        if (axis.gridCount == 0 || range <= 0) {
            return
        }
        val count = axis.gridCount + 1
        val rawInterval = range / count
        val interval = roundToOneSignificantFigure(rawInterval).toDouble()
        val first = ceil(start / interval) * interval
        axis.labelEntries = FloatArray(count + 1)
        var f: Double = first
        var i = 0
        while (i < count + 1) {
            axis.labelEntries[i] = f.toFloat()
            f += interval
            ++i
        }
    }

    /**
     * Rounds the given number to the given number of significant digits. Based on an answer on
     * [Stack Overflow](http://stackoverflow.com/questions/202302).
     */
    private fun roundToOneSignificantFigure(num: Double): Float {
        val d = ceil(log10(if (num < 0) -num else num).toFloat().toDouble()).toFloat()
        val power = 1 - d.toInt()
        val magnitude = 10.0.pow(power.toDouble()).toFloat()
        val shifted = (num * magnitude).roundToInt()
        return shifted / magnitude
    }

    private fun computeAxisStopsY(axis: AxisY) {
        val min = axis.yMin.toDouble()
        val max = axis.yMax.toDouble()
        val count = axis.gridCount + 1
        val interval = (max - min) / count
        axis.labelEntries = FloatArray(count + 1)
        if (min < Float.MAX_VALUE && max > -Float.MAX_VALUE && min <= max) {
            var f = min
            var j = 0
            while (j < count + 1) {
                axis.labelEntries[j] = f.toFloat()
                f += interval
                j++
            }
        }
    }

    /**
     * 画坐标轴文本
     */
    private fun drawAxisLabels(canvas: Canvas) {
        if (!axis.isLabelEnable) return
        if (axis.labels.isNullOrEmpty()) return
        val labels = axis.labels ?: emptyList()
        labelTextPaint.color = axis.labelTextColor
        labelTextPaint.textSize = axis.labelTextSize
        var x = 0f
        var y = 0f
        when (axis.axisPosition) {
            AxisX.TOP, AxisX.TOP_INSIDE -> {
                x = contentRect.left.toFloat()
                y = contentRect.top.toFloat()
            }

            AxisX.BOTTOM, AxisX.BOTTOM_INSIDE -> {
                x = contentRect.left.toFloat()
                y = (contentRect.top + contentRect.height()).toFloat()
            }

            AxisY.LEFT_INSIDE, AxisY.LEFT_OUTSIDE -> {
                x = contentRect.left.toFloat()
                y = contentRect.bottom.toFloat()
            }

            AxisY.RIGHT_INSIDE, AxisY.RIGHT_OUTSIDE -> {
                x = contentRect.right.toFloat()
                y = contentRect.bottom.toFloat()
            }
        }
        var labelOffset: Int
        var labelLength: Int
        if (axis is AxisX) {
            // X轴
            val width = contentRect.width() / labels.size.toFloat()
            for (i in labels.indices) {
                val labelCharArray = labels[i].toCharArray()
                labelLength = labelCharArray.size
                System.arraycopy(
                    labelCharArray,
                    0,
                    labelBuffer,
                    labelBuffer.size - labelLength,
                    labelLength
                )
                labelOffset = labelBuffer.size - labelLength
                labelTextPaint.textAlign = Paint.Align.CENTER
                val colorSetter = axis.labelColorSetter
                if (colorSetter != null) {
                    labelTextPaint.color = colorSetter.getColorByIndex(i)
                }
                canvas.drawText(
                    labelBuffer, labelOffset, labelLength,
                    width * 0.5f + getDrawX(i / labels.size.toFloat()),
                    y + labelTextPaint.textSize, labelTextPaint
                )
            }
        } else {
            // Y轴
            val height = contentRect.height() / (labels.size - 1f)
            var separation = 0f
            for (i in labels.indices) {
                val labelCharArray = labels[i].toCharArray()
                labelLength = labelCharArray.size
                System.arraycopy(
                    labelCharArray,
                    0,
                    labelBuffer,
                    labelBuffer.size - labelLength,
                    labelLength
                )
                labelOffset = labelBuffer.size - labelLength
                when (axis.axisPosition) {
                    AxisY.LEFT_OUTSIDE, AxisY.RIGHT_INSIDE -> {
                        labelTextPaint.textAlign = Paint.Align.RIGHT
                        separation = -axis.labelSeparation
                    }

                    AxisY.LEFT_INSIDE, AxisY.RIGHT_OUTSIDE -> {
                        labelTextPaint.textAlign = Paint.Align.LEFT
                        separation = axis.labelSeparation
                    }
                }
                val textHeightOffset = (labelTextPaint.descent() + labelTextPaint.ascent()) / 2
                val colorSetter = axis.labelColorSetter
                if (colorSetter != null) {
                    labelTextPaint.color = colorSetter.getColorByIndex(i)
                }
                canvas.drawText(
                    labelBuffer, labelOffset, labelLength,
                    x + separation,
                    y - i * height - textHeightOffset,
                    labelTextPaint
                )
            }
        }
    }

    private fun drawGridLabels(canvas: Canvas) {
        if (!axis.isLabelEnable) return
        val labels: FloatArray = axis.labelEntries
        if (labels.isEmpty()) return

        labelTextPaint.color = axis.labelTextColor
        labelTextPaint.textSize = axis.labelTextSize

        var x = 0f
        var y = 0f

        when (axis.axisPosition) {
            AxisX.TOP, AxisX.TOP_INSIDE -> {
                x = contentRect.left.toFloat()
                y = contentRect.top.toFloat()
            }

            AxisX.BOTTOM, AxisX.BOTTOM_INSIDE -> {
                x = contentRect.left.toFloat()
                y = (contentRect.top + contentRect.height()).toFloat()
            }

            AxisY.LEFT_INSIDE, AxisY.LEFT_OUTSIDE -> {
                x = contentRect.left.toFloat()
                y = contentRect.bottom.toFloat()
            }

            AxisY.RIGHT_INSIDE, AxisY.RIGHT_OUTSIDE -> {
                x = contentRect.right.toFloat()
                y = contentRect.bottom.toFloat()
            }
        }

        var labelOffset: Int
        var labelLength: Int

        if (axis is AxisX) { // X轴
            val width: Float = contentRect.width() / (labels.size - 1f)
            for (i in labels.indices) {
                val valueFormatter = axis.labelValueFormatter
                if (valueFormatter == null) {
                    labelLength = FloatUtils.formatFloatValue(labelBuffer, labels[i], 2)
                } else {
                    val labelCharArray = valueFormatter.format(labels[i], i).toCharArray()
                    labelLength = labelCharArray.size
                    System.arraycopy(labelCharArray, 0, labelBuffer, labelBuffer.size - labelLength, labelLength)
                }
                labelOffset = labelBuffer.size - labelLength

                when (i) {
                    0 -> {
                        labelTextPaint.textAlign = Paint.Align.LEFT
                    }
                    labels.size - 1 -> {
                        labelTextPaint.textAlign = Paint.Align.RIGHT
                    }
                    else -> {
                        labelTextPaint.textAlign = Paint.Align.CENTER
                    }
                }

                val colorSetter = axis.labelColorSetter
                if (colorSetter != null) {
                    labelTextPaint.color = colorSetter.getColorByIndex(i)
                }
                canvas.drawText(
                    labelBuffer, labelOffset, labelLength,
                    x + i * width,  // - textWidth * 0.5f,
                    y + labelTextPaint.textSize,
                    labelTextPaint
                )
            }
        } else {
            // Y轴
            val height: Float = contentRect.height() / (labels.size - 1f)
            var separation = 0f
            for (i in labels.indices) {
                val valueFormatter = axis.labelValueFormatter
                if (valueFormatter == null) {
                    labelLength = FloatUtils.formatFloatValue(labelBuffer, labels[i], 2)
                } else {
                    val labelCharArray = valueFormatter.format(labels[i], i).toCharArray()
                    labelLength = labelCharArray.size
                    System.arraycopy(labelCharArray, 0, labelBuffer, labelBuffer.size - labelLength, labelLength)
                }

                labelOffset = labelBuffer.size - labelLength

                when (axis.axisPosition) {
                    AxisY.LEFT_OUTSIDE, AxisY.RIGHT_INSIDE -> {
                        labelTextPaint.textAlign = Paint.Align.RIGHT
                        separation = -axis.labelSeparation
                    }

                    AxisY.LEFT_INSIDE, AxisY.RIGHT_OUTSIDE -> {
                        labelTextPaint.textAlign = Paint.Align.LEFT
                        separation = axis.labelSeparation
                    }
                }
                var textHeightOffset: Float = (labelTextPaint.descent() + labelTextPaint.ascent()) / 2

                if (i == 0) {
                    // Bottom
                    textHeightOffset = axis.labelSeparation
                } else if (i == labels.size - 1) {
                    // Top
                    textHeightOffset += textHeightOffset - axis.labelSeparation
                }

                val colorSetter = (axis as AxisY).labelColorSetter
                if (colorSetter != null) {
                    labelTextPaint.color = colorSetter.getColorByIndex(i)
                }
                canvas.drawText(
                    labelBuffer, labelOffset, labelLength,
                    x + separation,
                    y - i * height - textHeightOffset,
                    labelTextPaint
                )
            }
        }
    }

    /**
     * 画网格线
     */
    fun drawGridLines(canvas: Canvas) {
        val count = axis.gridCount + 1
        gridPaint.strokeWidth = axis.gridThickness
        gridPaint.color = axis.gridColor
        if (axis.dashedGridIntervals != null && axis.dashedGridPhase > 0) {
            gridPaint.pathEffect = DashPathEffect(axis.dashedGridIntervals, axis.dashedGridPhase)
        }
        if (axis is AxisX) {
            val width = contentRect.width() / count.toFloat()
            for (i in 1 until count) {
                if (axis.girdLineColorSetter != null) {
                    gridPaint.color = axis.girdLineColorSetter!!.getColorByIndex(axis.gridColor, i)
                }
                canvas.drawLine(
                    contentRect.left + i * width,
                    contentRect.top.toFloat(),
                    contentRect.left + i * width,
                    contentRect.bottom.toFloat(),
                    gridPaint
                )
            }
        }
        if (axis is AxisY) {
            val height = contentRect.height() / count.toFloat()
            for (i in 1 until count) {
                if (axis.girdLineColorSetter != null) {
                    gridPaint.color = axis.girdLineColorSetter!!.getColorByIndex(axis.gridColor, i)
                }
                canvas.drawLine(
                    contentRect.left.toFloat(),
                    contentRect.top + i * height,
                    contentRect.right.toFloat(),
                    contentRect.top + i * height,
                    gridPaint
                )
            }
        }
    }

    fun drawLabels(canvas: Canvas) {
        if (axis.labels == null) {
            drawGridLabels(canvas)
        } else {
            drawAxisLabels(canvas)
        }
    }
}
